daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/ledgerinteraction/GrpcErrorParser.scala

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf.engine.script.v2.ledgerinteraction
+
+import com.daml.lf.data.Ref._
+import com.daml.lf.transaction.{GlobalKey, TransactionVersion}
+import com.daml.lf.value.Value.ContractId
+import com.daml.lf.value.ValueCoder
+import com.daml.lf.value.ValueCoder.CidDecoder
+import com.google.protobuf.ByteString
+import io.grpc.StatusRuntimeException
+import scala.reflect.ClassTag
+
+object GrpcErrorParser {
+  val decodeValue = (s: String) =>
+    ValueCoder.decodeValue(CidDecoder, TransactionVersion.VDev, ByteString.copyFromUtf8(s)).toOption
+
+  val parseList = (s: String) => s.tail.init.split(", ").toSeq
+
+  // Converts a given SubmitError into a SubmitError. Wraps in an UnknownError if its not what we expect, wraps in a TruncatedError if we're missing resources
+  def convertStatusRuntimeException(s: StatusRuntimeException): SubmitError = {
+    import io.grpc.protobuf.StatusProto
+    import com.daml.error.utils.ErrorDetails._
+    import com.daml.error.ErrorResource
+
+    val grpcStatus = StatusProto.fromThrowable(s)
+    val details = from(grpcStatus)
+    val message = grpcStatus.getMessage()
+    val oErrorInfoDetail = details.collectFirst { case eid: ErrorInfoDetail => eid }
+    val errorCode = oErrorInfoDetail.fold("UNKNOWN")(_.errorCodeId)
+    val resourceDetails = details.collect { case rd: ResourceInfoDetail => rd }
+    val errorResourceFromString = ErrorResource.fromString _
+
+    def classNameOf[A: ClassTag]: String = implicitly[ClassTag[A]].runtimeClass.getSimpleName
+
+    // Builds an appropriate TruncatedError if the given partial function doesn't match
+    def caseErr[A <: SubmitError: ClassTag](
+        handler: PartialFunction[Seq[(ErrorResource, String)], A]
+    ): SubmitError =
+      handler
+        .lift(
+          resourceDetails.collect {
+            case ResourceInfoDetail(name, errorResourceFromString.unlift(res)) => (res, name)
+          }
+        )
+        .getOrElse(new SubmitError.TruncatedError(classNameOf[A], message))
+
+    errorCode match {
+      case "CONTRACT_NOT_FOUND" =>
+        caseErr {
+          case Seq((ErrorResource.ContractId, cid)) =>
+            SubmitError.ContractNotFound(ContractId.assertFromString(cid))
+          // TODO: what should we do here? Currently just taking the first contract id
+          case Seq((ErrorResource.ContractIds, cids)) =>
+            SubmitError.ContractNotFound(ContractId.assertFromString(parseList(cids).head))
```

Comment:
Unsure how often this can happen. I can update the error in daml to take a list of contract ids, but that'll be strange for the user. I could also add a multi variant of it, but then users will likely have to handle both.
I dont know what the user should expect here, nor do I know when we'd ever see this error with multiple ids (the engine doesn't throw it)

Reply:
I see it is produced in `ledger/ledger-api-errors/src/main/scala/com/daml/error/definitions/groups/ConsistencyErrors.scala`, under `MultipleContractsNotFound`, but not much beyond that.

Can we default to some NonEmpty list in the `ContractNotFound` error on the daml side, rather than adding a new variant?

Reply:
> Unsure how often this can happen. 
please ask on #team-canton slack or during the AMA

Reply:
I've reimplemented this as a NonEmpty for now, multiple contract errors are thrown only by canton sandbox, but its unclear
a) what that even is
b) how often it happens
We can reduce this back if it turns out that its never used, but for now we can be sure we're not throwing away information